### PR TITLE
feat: Host rustdoc of `master` on GitHub pages

### PR DESCRIPTION
**Pull request by umgefahren**
## Description

This PR hosts the documentation of the current master branch on GitHub Pages. The URL should be: https://libp2p.github.io/rust-libp2p

I tested the configuration on my repo and it works. You can see the result under: https://umgefahren.github.io/rust-libp2p

## Notes

This PR requires a change of settings in the pages tab. The screenshot below shows the correct configuration:

image

## Links to any relevant issues

Solving #3125. Sorry for opening so many issues and then immediately providing PRs.

## Open Questions

Is the community happy with this.

### CODE REVIEW

File: .github/workflows/docs.yml
Comment (thomaseizinger) [inline .github/workflows/docs.yml:22]: Would it make sense to add caching here and pin the version of the nightly compiler?

Comment (umgefahren) [reply .github/workflows/docs.yml:22]: Yes we could do that. But I don't see that much benefit here. Pinning would mean we have to change the pinned version every time a new Rust version is released (potentially) and since the workflow is only triggered when push to master happens the performance benefit is not that big. 

But sure I could pin

Comment (thomaseizinger) [reply .github/workflows/docs.yml:22]: Fair enough, let's not bother then!

**Review by thomaseizinger — APPROVED** [review]

**Comment by thomaseizinger** [comment]
LGTM!

I think the changes to the GitHub pages configuration will need to be made over here: https://github.com/libp2p/github-mgmt/blob/0e08ed66ddd586d3f31f98d1aa4cb54e3795737e/github/libp2p.yml#L3962

I got no idea what the correct config looks like. @galargh Can you help here?

**Comment by galargh** [comment]
> LGTM!
> 
> I think the changes to the GitHub pages configuration will need to be made over here: https://github.com/libp2p/github-mgmt/blob/0e08ed66ddd586d3f31f98d1aa4cb54e3795737e/github/libp2p.yml#L3962
> 
> I got no idea what the correct config looks like. @galargh Can you help here?

Sadly, terraform provider (and because of that - GitHub Management) doesn't support `GitHub Actions` as a Pages source yet - https://registry.terraform.io/providers/integrations/github/latest/docs/resources/repository#github-pages-configuration.

I made the required changes through the UI in this repo. You should be good to proceed.

**Comment by umgefahren** [comment]
What is missing here? It should be ready to merge from my POV

**Comment by mxinden** [comment]
~~Only waiting for a green CI at this point. Once green mergify should merge automatically.~~

Not sure why CI is failing. Will need to take a look.

**Comment by mxinden** [comment]
`libp2p-dcutr` and `libp2p-relay` changes seem unrelated. Discussion is here https://github.com/libp2p/rust-libp2p/pull/2647#issuecomment-1335189994. 

Moving forward here.

**Review by mxinden — APPROVED** [review]

**Comment by mxinden** [comment]
Docs page is fancy. Thanks @umgefahren.

**Comment by mxinden** [comment]
In case you want to draw more attention to it, referencing the page next to the already referenced libp2p docs in the main README.md would be a good idea.

**Pull request merged**
